src/main/java/com/shade/decima/ui/data/viewer/texture/controls/ImagePanel.java

```diff
@@ -8,8 +8,35 @@
 import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
 import java.awt.image.BufferedImage;
+import java.awt.image.FilteredImageSource;
+import java.awt.image.RGBImageFilter;
 import java.util.Objects;
 
+class ChannelFilter extends RGBImageFilter {
+    private String channel;
+
+    public ChannelFilter(String channel) {
+        this.channel = channel;
+        canFilterIndexColorModel = true;
+    }
+
+    public int filterRGB(int x, int y, int rgb) {
+        Color c = new Color(rgb, true);
+        int r = c.getRed();
+        int g = c.getGreen();
+        int b = c.getBlue();
+        int a = c.getAlpha();
+        return switch (this.channel) {
+            case "RGB" -> new Color(r, g, b, 255).getRGB();
+            case "R" -> new Color(r, r, r, 255).getRGB();
+            case "G" -> new Color(g, g, g, 255).getRGB();
+            case "B" -> new Color(b, b, b, 255).getRGB();
+            case "A" -> new Color(255, 255, 255, a).getRGB();
+            default -> throw new IllegalArgumentException("channel not in {'RGB', 'R', 'G', 'B', 'A'}");
+        };
```

Comment:
Fewer allocations result in better performance, yet you lack readability in return.

You can use bitwise operations to reduce allocations. To retrieve individual components:

```java
final int a = (rgb >> 24) & 0xff; 
final int r = (rgb >> 16) & 0xff; 
final int g = (rgb >> 8) & 0xff; 
final int b = (rgb >> 0) & 0xff; 
```

To combine them back, e.g. to fill just all channels with red:

```java
return r << 16 | r << 8 | r;
```

Reply:
May I use your RGB utils from src/main/java/com/shade/decima/ui/data/viewer/texture/util/RGB.java for better readability

Reply:
This method is called for every pixel on every update, which is A LOT. So even if you use the `RGB` class, It will introduce allocations anyways, so I'd stick with the bitwise version.

I'll be happy to assist if you need help.